pandas/core/internals/blocks.py

```diff
@@ -839,6 +839,9 @@ def setitem(self, indexer, value):
 
         # coerce if block dtype can store value
         values = self.values
+        if is_object_dtype(values.dtype) and isinstance(value, DatetimeArray):
```

Comment:
Why is `values.dtype` object dtype here? That surprises me.

Reply:
When a `datetime[ns, tz]` column is assigned to a new column using an index an empty column is created with fill values, for any `datetime` or `timedelta` type the fill value is `NaT` (`pandas.core.dtypes.missing._infer_fill_value`) with the corresponding dtype. For `datetime[ns, tz]` the fill value is still of dtype `datetime[ns]`. In `Block.setitem` the common dtype between `datetime[ns]` and `datetime[ns, tz]` is `object` and the block is converted to an `ObjectBlock` and its `setitem` method is called which leads to the `values.dtype` to be `object`.  You could change this behaviour for the specific case of `datetime[ns, tz]` by changing the fill value but what would the fill value be for creating an empty `datetime[ns, tz]` column?

Reply:
>  In Block.setitem the common dtype between datetime[ns] and datetime[ns, tz] is object and the block is converted to an ObjectBlock and its setitem method is called which leads to the values.dtype to be object

I think this is the issue. We should be able to go from `datetime[ns] -> datetime[ns, tz]` without going through object dtype. Fixing it earlier to fill NaT with datetime64[ns, tz] dtype would also be an option. I'm not sure which is preferable.

Reply:
`NaT` itself has no timezone information so a `datetime[ns, tz]` `NaT` object cannot be created, the current fill code uses a numpy `datetime64` object which does not support timezone information. You could wrap the `NaT` in a `DatetimeIndex` to create a `datetime[ns, tz]` object but I do no think that is a good solution. I think that one solution going forward would be to support a  `datetime[ns, tz]` `NaT` object but that should be a separate discussion.

Reply:
> I think that one solution going forward would be to support a datetime[ns, tz] NaT object but that should be a separate discussion.

thats not unreasonable, but its also very unlikely, especially before 2.0

> You could wrap the NaT in a DatetimeIndex to create a datetime[ns, tz] object but I do not think that is a good solution

This sounds reasonable, can you expand on why you think it isnt a good solution?

Reply:
> This sounds reasonable, can you expand on why you think it isnt a good solution?

The missing value is returned by `pandas.core.dtypes.missing._infer_fill_value`, the `missing` module is used in the definition of `DatetimeIndex` so you end up with a fix to the indexer `_setitem_with_indexer` looking something like this -
```
if hasattr(value, "dtype") and is_datetime64tz_dtype(value.dtype):
    self.obj[key] = DatetimeIndex(["NaT"] * len(value), tz=value.dtype.tz)
else:
    self.obj[key] = _infer_fill_value(value)
```
To me this does not seem to be a good way to fix the problem, it does however retain the dtype so if there is a consensus that this is the optimal solution for now I can add a commit with this modification.

Reply:
bit of a kludge, but what if we made _infer_fill_value return a length=1 DatetimeArray in this case?

Reply:
`_infer_fill_value`  is in the `pandas.core.dtypes.missing` module, the `isna` function from that module is used in the definition of `DatetimeArray`. `_infer_fill_value` is only used in the `_setitem_with_indexer` function so this could be implemented by moving the `_infer_fill_value` function to the `pandas.core.indexing` module. The code would be quite similar - you would have to create a `DatetimeArray` the size of `value` for the assignment to work.

Reply:
why are you doing this in blocks generically?

this should only be in an Object block.

Reply:
@jreback yes, I can move this, I was waiting for an update on https://github.com/pandas-dev/pandas/pull/32479#discussion_r392056376 and https://github.com/pandas-dev/pandas/pull/32479#discussion_r395987552 alternate solutions that do not require this code. Also relevant for your observation https://github.com/pandas-dev/pandas/pull/32479#discussion_r399480077

Reply:
There is now a different fix applied, a slice assignment to a new column will still create an `object` `dtype` for most extension array types.